### ISSUE TITLE
Show tooltip when mouse inside a cell instead of an item

### ISSUE THREAD

**Issue by SuperWangKai** [ocornut/imgui#6347]
Hi Omar,

Thanks again for the amazing UI solution!
I'm using Table APIs and trying to to show a tooltip for each of the cells, however, if using `ImGui::IsItemHovered()` as the condition, the hovering rect is the actual item rect and if the item is very small, e.g. one character, it's very hard to have the tooltip showed. 
Is there a way to trigger the tooltip in the range of a cell size?

**Screen record**
![table_tooltip](https://user-images.githubusercontent.com/15721569/232842452-5f55e69c-54d2-418d-af78-5c8017209a11.gif)

**Version/Branch of Dear ImGui:**

```
Dear ImGui 1.89.4 WIP (18931)
--------------------------------

sizeof(size_t): 8, sizeof(ImDrawIdx): 2, sizeof(ImDrawVert): 20
define: __cplusplus=199711
define: _WIN32
define: _WIN64
define: _MSC_VER=1935
define: _MSVC_LANG=201703
define: IMGUI_HAS_VIEWPORT
define: IMGUI_HAS_DOCK
--------------------------------

io.BackendPlatformName: imgui_impl_win32
io.BackendRendererName: imgui_impl_dx12
io.ConfigFlags: 0x00000443
 NavEnableKeyboard
 NavEnableGamepad
 DockingEnable
 ViewportsEnable
io.ConfigViewportsNoDecoration
io.ConfigInputTextCursorBlink
io.ConfigWindowsResizeFromEdges
io.ConfigMemoryCompactTimer = 60.0
io.BackendFlags: 0x00001C0E
 HasMouseCursors
 HasSetMousePos
 PlatformHasViewports
 HasMouseHoveredViewport
 RendererHasVtxOffset
 RendererHasViewports
--------------------------------

io.Fonts: 3 fonts, Flags: 0x00000000, TexSize: 2048,2048
io.DisplaySize: 700.00,900.00
io.DisplayFramebufferScale: 1.00,1.00
--------------------------------

style.WindowPadding: 12.00,12.00
style.WindowBorderSize: 1.00
style.FramePadding: 6.00,4.00
style.FrameRounding: 0.00
style.FrameBorderSize: 0.00
style.ItemSpacing: 12.00,6.00
style.ItemInnerSpacing: 6.00,6.00
```

**Code to reproduce the issue**

```c++
constexpr ImGuiTableFlags flags = ImGuiTableFlags_Resizable | ImGuiTableFlags_Borders;
const float TEXT_BASE_HEIGHT = ImGui::GetTextLineHeightWithSpacing();

// Test tooltip.
if (ImGui::TreeNodeEx("Properties", ImGuiTreeNodeFlags_SpanFullWidth | ImGuiTreeNodeFlags_DefaultOpen))
{
    if (ImGui::BeginTable("prop", 2, flags, ImVec2(0.0f, TEXT_BASE_HEIGHT * 10)))
    {
        ImGui::TableSetupScrollFreeze(0, 1);
        ImGui::TableSetupColumn("Property");
        ImGui::TableSetupColumn("Value", ImGuiTableColumnFlags_WidthStretch);
        // ImGui::TableHeadersRow();

        for (uint32_t i = 0; i < 10; ++i)
        {
            // Prop 1
            ImGui::TableNextRow();
            ImGui::TableNextColumn();
            ImGui::Text("Name_%u", i);

            // Tooltip for Prop 1
            if (ImGui::IsItemHovered())
            {
                ImGui::BeginTooltip();
                ImGui::Text("Tooltip on name %u", i);
                ImGui::EndTooltip();
            }
            // Value 1
            ImGui::TableNextColumn();
            ImGui::Text("Value_%u", i);

            // Tooltip for Value 1
            if (ImGui::IsItemHovered())
            {
                ImGui::BeginTooltip();
                ImGui::Text("Tooltip on value %u", i);
                ImGui::EndTooltip();
            }
        }

        ImGui::EndTable();
    }

    ImGui::TreePop();
}
```

Best regards,
Kai

**Comment by ocornut** [comment]
This seems like the same as https://github.com/ocornut/imgui/issues/6250#issuecomment-1479767259

TL;DR; There's no "one liner" way of telling is a cell is hovered in every situation because we can't always reliably get line height ahead of time. Essentially we are missing a `TableGetHoveredRow()` function. But if you call `TableNextRow()` with a height and don't step beyond this height, then you can tell the Y1/Y2 of your row.

Use `TableGetHoveredColumn()` + look at the code from `TableGetCellBgRect()` to find Y1/Y2 and you can then tell if a cell is hovered if you pass your row height ahead of time in `TableNextRow()`

**Comment by SuperWangKai** [comment]
Hi Omar,

Thank you so much for the help.
I guess due to the fact that Text is used instead of a hoverable item? `TableGetHoveredColumn` always returns `-1` here.

I'm now using `column->ClipRect.Min` to tell the column. Seems working here.
```c++
constexpr ImGuiTableFlags flags = ImGuiTableFlags_Resizable | ImGuiTableFlags_Borders;
const float TEXT_BASE_HEIGHT = ImGui::GetTextLineHeightWithSpacing();

// Test tooltip.
if (ImGui::TreeNodeEx("Properties", ImGuiTreeNodeFlags_SpanFullWidth | ImGuiTreeNodeFlags_DefaultOpen))
{
    if (ImGui::BeginTable("prop", 2, flags, ImVec2(0.0f, TEXT_BASE_HEIGHT * 10)))
    {
        ImGui::TableSetupScrollFreeze(0, 1);
        ImGui::TableSetupColumn("Property");
        ImGui::TableSetupColumn("Value", ImGuiTableColumnFlags_WidthStretch);
        // ImGui::TableHeadersRow();

        auto mouse_pos = ImGui::GetMousePos();

        for (uint32_t i = 0; i < 10; ++i)
        {
            // Prop 1
            ImGui::TableNextRow(0, TEXT_BASE_HEIGHT); // WA for knowing the height of the row.
            ImGui::TableNextColumn();
            ImGui::Text("Name_%u", i);

            auto table = ImGui::GetCurrentTable();
            assert(table);

            //auto hovered_col = ImGui::TableGetHoveredColumn(); // hovered_col always equals -1

            ImGuiTableColumn* column = &table->Columns[0];

            // Tooltip for Prop1
            if ((mouse_pos.x > column->ClipRect.Min.x)
                && (mouse_pos.x < column->ClipRect.Max.x)
                && (mouse_pos.y > table->RowPosY1)
                && (mouse_pos.y < table->RowPosY2))
            {
                ImGui::BeginTooltip();
                ImGui::Text("Tooltip on name %u", i);
                ImGui::EndTooltip();
            }
            // Value 1
            ImGui::TableNextColumn();
            ImGui::Text("Value_%u", i);

            column = &table->Columns[1];

            // Tooltip for Value 1
            if ((mouse_pos.x > column->ClipRect.Min.x)
                && (mouse_pos.x < column->ClipRect.Max.x)
                && (mouse_pos.y > table->RowPosY1)
                && (mouse_pos.y < table->RowPosY2))
            {
                ImGui::BeginTooltip();
                ImGui::Text("Tooltip on value %u", i);
                ImGui::EndTooltip();
            }
        }

        ImGui::EndTable();
    }

    ImGui::TreePop();
}
```

**Comment by ocornut** [comment]
> TableGetHoveredColumn always returns -1 here.

That's a problem which should be investigated.
If i look at the code it looks correct.
In TableUpdateLayout(): 
```cpp
        // Detect hovered column
        if (is_hovering_table && g.IO.MousePos.x >= column->ClipRect.Min.x && g.IO.MousePos.x < column->ClipRect.Max.x)
            table->HoveredColumnBody = (ImGuiTableColumnIdx)column_n;
```

Can you check the value in Metrics?
![image](https://user-images.githubusercontent.com/8225057/233041393-1c22cc6c-3997-4662-9105-26905edf6b0a.png)

```cpp
            ImGuiTableColumn* column = &table->Columns[0];

            // Tooltip for Prop1
            if ((mouse_pos.x > column->ClipRect.Min.x)
                && (mouse_pos.x < column->ClipRect.Max.x)
                && (mouse_pos.y > table->RowPosY1)
                && (mouse_pos.y < table->RowPosY2))
            {
```
At minimum you should wrap this in a function, but I suspect you might as well simply use `TableGetCellBgRect(g.CurrentTable, g.CurrentTable->CurrentColumn).Contains(mouse_pos);`

My suggestion to use `TableGetHoveredColumn()` was to call a simplified version and only on column is hovered.

**Comment by SuperWangKai** [comment]
You are right. It's must too late yesterday :)

Both `TableGetHoveredColumn` and `TableGetCellBgRect(g.CurrentTable, g.CurrentTable->CurrentColumn).Contains(mouse_pos);` work.

So here is the code:
```c++
constexpr ImGuiTableFlags flags = ImGuiTableFlags_Resizable | ImGuiTableFlags_Borders;
const float TEXT_BASE_HEIGHT = ImGui::GetTextLineHeightWithSpacing();

// Test tooltip.
if (ImGui::TreeNodeEx("Properties", ImGuiTreeNodeFlags_SpanFullWidth | ImGuiTreeNodeFlags_DefaultOpen))
{
    if (ImGui::BeginTable("prop", 2, flags, ImVec2(0.0f, TEXT_BASE_HEIGHT * 10)))
    {
        ImGui::TableSetupScrollFreeze(0, 1);
        ImGui::TableSetupColumn("Property");
        ImGui::TableSetupColumn("Value", ImGuiTableColumnFlags_WidthStretch);
        // ImGui::TableHeadersRow();

        // Function returns if current cell is hovered.
        auto cell_hovered = []()
        {
            auto mouse_pos = ImGui::GetMousePos();
            auto table = ImGui::GetCurrentTable();

            return ImGui::TableGetCellBgRect(table, table->CurrentColumn).Contains(mouse_pos);
        };

        for (uint32_t i = 0; i < 10; ++i)
        {
            // Prop 1
            ImGui::TableNextRow(0, TEXT_BASE_HEIGHT);  // WA for knowing the height of the row.
            ImGui::TableNextColumn();
            ImGui::Text("Name_%u", i);

            // Tooltip for Prop 1
            if (cell_hovered())
            {
                ImGui::BeginTooltip();
                ImGui::Text("Tooltip on Name %u", i);
                ImGui::EndTooltip();
            }
            // Value 1
            ImGui::TableNextColumn();
            ImGui::Text("Value_%u", i);

            // Tooltip for Value 1
            if (cell_hovered())
            {
                ImGui::BeginTooltip();
                ImGui::Text("Tooltip on Value %u", i);
                ImGui::EndTooltip();
            }
        }

        ImGui::EndTable();
    }

    ImGui::TreePop();
}
```

**Comment by SuperWangKai** [comment]
So here comes the next question... how to handle the delay for the tooltip.
Here https://github.com/ocornut/imgui/issues/1485 we have `GImGui->HoveredIdTimer > MY_TIME_THRESHOLD` works great, I wonder how to simulate one in our cell case?

**Comment by ocornut** [comment]
Closing this.
My suggestion to use `TableGetHoveredColumn()` was only to make code a little bit faster:

```
if (column_hovered == 0 && row_hovered(0))
```
Where `row_hovered` can only do the Y1/Y2 check, but your solution seems to work.

I'll post here and #6250 if I decide to add a `TableGetHoveredRow()` or `TableIsHoveredCell()` etc.

**Comment by mrawlingst** [comment]
Sorry for the necro bump, but I came across this thread and found @devkaiwang's solution really helpful as inspiration. I wanted to share my own approach in case it helps someone else down the line.

This is in C# - I'm lazy to re-implement in C++ but it should be simple enough to translate it.

Note: I'm using [Hexa.NET.ImGui - v2.2.7](https://github.com/HexaEngine/Hexa.NET.ImGui) which is more updated than ImGuiNET.

```csharp
public unsafe static bool IsTableCellHovered(Vector2 cellMin, bool ignorePopups = false)
{
    var cellWidth = ImGui.GetColumnWidth();
    var cellMax = new Vector2(cellMin.X + cellWidth, cellMin.Y + ImGui.GetTextLineHeightWithSpacing());
    var mousePos = ImGui.GetMousePos();

    var clipRect = ImGui.GetWindowDrawList().ClipRectStack.Back;
    var clipMin = new Vector2(clipRect->X, clipRect->Y);
    var clipMax = new Vector2(clipRect->Z, clipRect->W);

    var IsVisible = cellMax.X > clipMin.X && cellMin.X < clipMax.X &&
                    cellMax.Y > clipMin.Y && cellMin.Y < clipMax.Y;

    return IsVisible && mousePos.X >= cellMin.X && mousePos.X <= cellMax.X &&
            mousePos.Y >= cellMin.Y && mousePos.Y <= cellMax.Y &&
            (ignorePopups || !IsAnyPopupOpen());
}
```

You need to pass a Vector2 `cellMin` which is set between `ImGui.TableNextColumn()` and content you add (e.g. `ImGui.Text()`).

`ignorePopups` - if this wasn't implemented, it would show the tooltips behind context menu if it was overlapping tables when hovering.
`IsAnyPopupOpen()` - you have to register popups and just iterate over a containers of Ids and check against `ImGui.IsPopupOpen()`

Popup Manager:
```csharp
private static HashSet<string> _registeredPopupIds = [];

public static void RegisterPopup(string pId)
{
    _registeredPopupIds.Add(pId);
}

public static void UnregisterPopup(string pId)
{
    _registeredPopupIds.Remove(pId);
}

public static bool IsAnyPopupOpen()
{
    foreach (var id in _registeredPopupIds)
    {
        if (ImGui.IsPopupOpen(id))
            return true;
    }
    return false;
}
```

You could probably implement a better solution for popup checking.

Example:
```csharp
var popupId = "popupId1";
RegisterPopup(popupId);

if (ImGui.BeginTable("ExampleTable", 2, ImGuiTableFlags.Borders | ImGuiTableFlags.RowBg))
{
    ImGui.TableNExtRow();

    var cellMin = Vector2.Zero;

    // First cell
    ImGui.TableNextColumn();
    cellMin = ImGui.GetCursorScreenPos(); // This must be before any content you add because it gets the beginning of the cell
    ImGui.Text("Cell Test");
    if (IsTableCellHovered(cellMin))
    {
        // Do things such as ImGui.BeginTooltip(); ... ImGui.EndTooltip();
    }

    // Second cell
    ImGui.TableNextColumn();
    cellMin = ImGui.GetCursorScreenPos();
    ImGui.PushId(popupId);
    ImGui.Text("Cell Test");
    ImGui.PopId();
    if (IsTableCellHovered(cellMin))
    {
        // Do things such as ImGui.BeginTooltip(); ... ImGui.EndTooltip();
    }
    if (ImGui.BeginPopupContextItem(popupId))
    {
        // Do things for your context menu
    }
}
```

This is a very crude solution, in my opinion, but this is the best I got. I would absolutely love an actual `IsTableCellHovered()` within the official API.

I'm using a table to manage the inventory. The table can be quite wide, and often the text doesn't fill the entire cell, leaving a lot of unused space. I wanted to make it so that hovering anywhere within the cell—rather than just over the text—shows a tooltip.

I think it absolutely can be improved but I'm not an expert with the ImGui API yet. So if anyone has a better suggestion even if it's in C++, please share!
